myia/abstract/infer.py

```diff
@@ -0,0 +1,679 @@
+        return self.constructors[g]
+
+    @_get_inferrer_for.register
+    def _get_inferrer_for(self, part: PartialApplication):
+        return PartialInferrer(
+            self.get_inferrer_for(part.fn),
+            part.args
+        )
+
+    @_get_inferrer_for.register
+    def _get_inferrer_for(self, j: JTransformedFunction):
+        return JInferrer(
+            self.get_inferrer_for(j.fn),
+            j.fn
+        )
+
+    @_get_inferrer_for.register
+    def _get_inferrer_for(self, vf: (VirtualFunction, TypedPrimitive)):
+        return VirtualInferrer(
+            vf.args,
+            vf.output
+        )
+
+    @_get_inferrer_for.register
+    def _get_inferrer_for(self, df: DummyFunction):
+        raise MyiaTypeError(f'Trying to call dummy')
+
+    @_get_inferrer_for.register
+    def _get_inferrer_for(self, mg: MetaGraphFunction):
+        if mg not in self.constructors:
+            self.constructors[mg] = MetaGraphInferrer(mg.metagraph)
+        return self.constructors[mg]
+
+    def get_inferrer_for(self, fn):
+        """Return the Inferrer for the given function."""
+        tracking = getattr(fn, 'tracking_id', None)
+        if tracking is None:
+            return self._get_inferrer_for(fn)
+        if fn not in self.constructors:
+            fn_generic = dc_replace(fn, tracking_id=None)
+            inf = self._get_inferrer_for(fn_generic)
+            self.constructors[fn] = TrackedInferrer(inf)
+        return self.constructors[fn]
+
+    async def execute(self, fn, *args):
+        """Infer the result of fn(*args)."""
+        infs = [self.get_inferrer_for(poss)
+                for poss in await fn.get()]
+        argrefs = [VirtualReference(a) for a in args]
+        return await execute_inferrers(self, infs, None, argrefs)
+
+    async def infer_apply(self, ref):
+        """Infer the type of a ref of an Apply node."""
+        ctx = ref.context
+        n_fn, *n_args = ref.node.inputs
+        # We await on the function node to get the inferrer
+        fn_ref = self.ref(n_fn, ctx)
+        fn = await fn_ref.get()
+        argrefs = [self.ref(node, ctx) for node in n_args]
+
+        if not isinstance(fn, AbstractFunction):
+            raise MyiaTypeError(f'Not a function: {fn}', refs=[fn_ref])
+
+        infs = [self.get_inferrer_for(poss)
+                for poss in await fn.get()]
+
+        return await self.loop.schedule(
+            execute_inferrers(self, infs, ref, argrefs),
+            context_map={
+                infer_trace: {**infer_trace.get(), ctx: ref}
+            }
+        )
+
+    async def infer_constant(self, ctref):
+        """Infer the type of a ref of a Constant node."""
+        v = self.pipeline.resources.convert(ctref.node.value)
+        res = to_abstract(v, ctref.context, ref=ctref)
+        t = build_type(res)
+        if dtype.ismyiatype(t, dtype.Number):
+            prio = 1 if dtype.ismyiatype(t, dtype.Float) else 0
```

Comment:
Do you have tests of 32 vs 64 bits for this?

Should we extend the priorities into an "upcast map"?

This should be done in a separate PR if we decide to do it.

Reply:
Realistically this is only going to be triggered for Python int/float (64 bits), because that's what a literal in the code is going to be, but I suppose it may be useful to cover other int/float types just to be safe.

I don't know about upcasting, we can discuss that separately if you wish.